File: problems/3901-4000/3944_remainder-cadence-ii/solution.cpp
#include <bits/stdc++.h>
using namespace std;
class Solution {
  public:
    long long alignRemainderRing(vector<int> &a, int k) {
        auto costs = [&](int p) {
            vector<long long> c(k), t(3 * k), pc(3 * k + 1), ps(3 * k + 1), o(k);
            for (int i = p; i < a.size(); i += 2)
                c[a[i] % k]++;
            for (int i = 0; i < 3 * k; i++) {
                t[i] = c[i % k];
                pc[i + 1] = pc[i] + t[i];
                ps[i + 1] = ps[i] + t[i] * i;
            }
            int h = k / 2;
            for (int x = 0; x < k; x++) {
                int m = x + k, l = m - h, r = m + k - 1 - h;
                long long lc = pc[m + 1] - pc[l], ls = ps[m + 1] - ps[l], rc = pc[r + 1] - pc[m + 1],
                          rs = ps[r + 1] - ps[m + 1];
                o[x] = m * lc - ls + rs - m * rc;
            }
            return o;
        };
        auto e = costs(0), o = costs(1);
        int p = 0, q = 1;
        if (o[q] < o[p])
            swap(p, q);
        for (int i = 2; i < k; i++)
            if (o[i] < o[p]) {
                q = p;
                p = i;
            } else if (o[i] < o[q])
                q = i;
        long long ans = LLONG_MAX;
        for (int x = 0; x < k; x++)
            ans = min(ans, e[x] + o[p == x ? q : p]);
        return ans;
    }
};

File: problems/3901-4000/3944_remainder-cadence-ii/solution.java
import java.util.*;

class Solution {

    public long alignRemainderRing(int[] a, int k) {
        long[] e = cost(a, k, 0),
            o = cost(a, k, 1);
        Integer[] ix = new Integer[k];
        for (int i = 0; i < k; i++) ix[i] = i;
        Arrays.sort(ix, (x, y) -> Long.compare(o[x], o[y]));
        long ans = Long.MAX_VALUE;
        for (int x = 0; x < k; x++) ans = Math.min(ans, e[x] + o[ix[0] == x ? ix[1] : ix[0]]);
        return ans;
    }

    long[] cost(int[] a, int k, int p) {
        long[] c = new long[k],
            pc = new long[3 * k + 1],
            ps = new long[3 * k + 1],
            out = new long[k];
        for (int i = p; i < a.length; i += 2) c[a[i] % k]++;
        for (int i = 0; i < 3 * k; i++) {
            pc[i + 1] = pc[i] + c[i % k];
            ps[i + 1] = ps[i] + c[i % k] * i;
        }
        int h = k / 2;
        for (int x = 0; x < k; x++) {
            int m = x + k,
                l = m - h,
                r = m + k - 1 - h;
            long lc = pc[m + 1] - pc[l],
                ls = ps[m + 1] - ps[l],
                rc = pc[r + 1] - pc[m + 1],
                rs = ps[r + 1] - ps[m + 1];
            out[x] = m * lc - ls + rs - m * rc;
        }
        return out;
    }
}
